### PR TITLE
Dev add i2s stm32h5

### PR DESCRIPTION
**Pull request by marwaiehm-st**
Add the STM32H5 family support for the test I2s_speed.
The driver is updated to handle correctly the commands TRIGGER_DRAIN.

### CODE REVIEW

**Comment by erwango** [comment]
@jeppenodgaard ^^

File: drivers/i2s/i2s_ll_stm32.c [hunk 1]
```diff
@@ -597,12 +612,35 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
...
 		LOG_ERR("TX error detected");
 		goto tx_disable;
 	}
 
+	/* check if the tx stop command can be handled */
+	/*if True, the tx stop command can be handled  */
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		/* check if all tx samples have been completely handled */
+		/*As stated in zephyr i2s specification, in case of DRAIN command */
+		/* Send all data in the transmit queue and stop the transmission.*/
```
Comment (jeppenodgaard) [inline drivers/i2s/i2s_ll_stm32.c:629]: I think multiline comment should follow linux' guidelines:
https://www.kernel.org/doc/html/latest/process/coding-style.html#commenting
This is also the case for other changes.

File: drivers/i2s/i2s_ll_stm32.c [hunk 2]
```diff
@@ -597,12 +612,35 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
...
 	if (stream->state == I2S_STATE_ERROR) {
 		LOG_ERR("TX error detected");
 		goto tx_disable;
 	}
 
+	/* check if the tx stop command can be handled */
+	/*if True, the tx stop command can be handled  */
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		/* check if all tx samples have been completely handled */
+		/*As stated in zephyr i2s specification, in case of DRAIN command */
```
Comment (jeppenodgaard) [inline drivers/i2s/i2s_ll_stm32.c:628]: Missing whitespace after * . Same in other changes.

**Review by jeppenodgaard — APPROVED** [review]

File: dts/arm/st/h5/stm32h5.dtsi [hunk 3]
```diff
@@ -491,6 +491,54 @@
...
 
+		i2s1: i2s@40013000 {
+			compatible = "st,stm32h7-i2s", "st,stm32-i2s";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40013000 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,
+				<&rcc STM32_SRC_PLL1_Q SPI1_SEL(0)>;
+			dmas = <&gpdma1 0 7 STM32_DMA_PERIPH_TX
+				&gpdma1 1 6 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_INC | \
```
Comment (erwango) [inline dts/arm/st/h5/stm32h5.dtsi:502]: `STM32_DMA_MEM_INC` is included in `STM32_DMA_PERIPH_RX`.

File: dts/arm/st/h5/stm32h5.dtsi [hunk 4]
```diff
@@ -491,6 +491,54 @@
...
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
```
Comment (erwango) [inline dts/arm/st/h5/stm32h5.dtsi:503]: Should we use `STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS` in TX as well ?
If not, let's add a comment to explain this configuration.

Nit: A new flag `STM32_DMA_16BITS` could be added as a short to `STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS`

File: dts/arm/st/h5/stm32h5.dtsi [hunk 5]
```diff
@@ -491,6 +491,54 @@
...
+			compatible = "st,stm32h7-i2s", "st,stm32-i2s";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40003800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00004000>,
+				<&rcc STM32_SRC_PLL1_Q SPI2_SEL(0)>;
+			dmas = <&gpdma1 2 9 (STM32_DMA_PERIPH_TX | STM32_DMA_MEM_INC | \
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
+					STM32_DMA_PRIORITY_HIGH)
+				&gpdma1 3 8 STM32_DMA_PERIPH_RX>;
```
Comment (erwango) [inline dts/arm/st/h5/stm32h5.dtsi:520]: On i2s1 configuration is the opposite:
RX: 16 Bits, prio high

File: dts/arm/st/h5/stm32h5.dtsi [hunk 6]
```diff
@@ -491,6 +491,54 @@
...
+			compatible = "st,stm32h7-i2s", "st,stm32-i2s";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40003c00 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00008000>,
+				<&rcc STM32_SRC_PLL1_Q SPI3_SEL(0)>;
+			dmas = <&gpdma1 4 11 (STM32_DMA_PERIPH_TX  | STM32_DMA_MEM_INC | \
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
+					STM32_DMA_PRIORITY_HIGH)
+				&gpdma1 5 10 STM32_DMA_PERIPH_RX>;
```
Comment (erwango) [inline dts/arm/st/h5/stm32h5.dtsi:536]: Ditto

File: dts/arm/st/h5/stm32h5.dtsi [hunk 7]
```diff
@@ -491,6 +491,54 @@
...
+			compatible = "st,stm32h7-i2s", "st,stm32-i2s";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40013000 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,
+				<&rcc STM32_SRC_PLL1_Q SPI1_SEL(0)>;
+			dmas = <&gpdma1 0 7 STM32_DMA_PERIPH_TX
+				&gpdma1 1 6 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_INC | \
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
+					STM32_DMA_PRIORITY_HIGH)>;
```
Comment (erwango) [inline dts/arm/st/h5/stm32h5.dtsi:504]: Is that required only for RX ?

File: tests/drivers/i2s/i2s_speed/boards/stm32h573i_dk.overlay [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+/* DMA Tx interrupt 0 has a higher Prio than Rx Dma interrupt 1 */
+&gpdma1 {
+	interrupts = <27 1 28 2 29 1 30 1 31 1 32 1 33 1 34 1>;
+	status = "okay";
+};
+
+&gpdma2 {
+	interrupts = <90 1 91 1 92 1 93 1 94 1 95 1 96 1 97 1>;
+	status = "okay";
+};
```
Comment (erwango) [inline tests/drivers/i2s/i2s_speed/boards/stm32h573i_dk.overlay:31]: Is that required ?

File: drivers/i2s/i2s_ll_stm32.c [hunk 9]
```diff
@@ -174,9 +174,6 @@ static int i2s_stm32_configure(const struct device *dev, enum i2s_dir dir,
 	int ret;
 
 	if (dir == I2S_DIR_RX) {
-#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_i2s)
-		return -ENOSYS;
-#endif
```
Comment (erwango) [inline drivers/i2s/i2s_ll_stm32.c:179]: Please make it a dedicated commit.

File: drivers/i2s/i2s_ll_stm32.c [hunk 10]
```diff
@@ -652,11 +700,6 @@ static uint32_t i2s_stm32_irq_udr_count;
 static void i2s_stm32_isr(const struct device *dev)
 {
 	const struct i2s_stm32_cfg *cfg = dev->config;
-	struct i2s_stm32_data *const dev_data = dev->data;
-	struct stream *stream = &dev_data->rx;
-
-	LOG_ERR("%s: err=%d", __func__, (int)LL_I2S_ReadReg(cfg->i2s, SR));
-	stream->state = I2S_STATE_ERROR;
```
Comment (erwango) [inline drivers/i2s/i2s_ll_stm32.c:659]: A dedicated commit would be required.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/dma/dma_stm32u5.c [hunk 11]
```diff
@@ -544,11 +544,9 @@ static int dma_stm32_reload(const struct device *dev, uint32_t id,
...
 
 	if (stream->source_periph) {
-		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id),
-				     size / stream->src_size);
+		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id), size);
 	} else {
-		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id),
-				     size / stream->dst_size);
+		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id), size);
 	}
```
Comment (mathieuchopstm) [inline drivers/dma/dma_stm32u5.c:550]: Identical code in `if` and `else` blocks; check should be removed.

File: drivers/i2s/i2s_ll_stm32.c (hunk 9 above)
Comment (marwaiehm-st) [reply drivers/i2s/i2s_ll_stm32.c:179]: Done

File: drivers/i2s/i2s_ll_stm32.c (hunk 10 above)
Comment (marwaiehm-st) [reply drivers/i2s/i2s_ll_stm32.c:659]: Done

File: drivers/dma/dma_stm32u5.c (hunk 11 above)
Comment (marwaiehm-st) [reply drivers/dma/dma_stm32u5.c]: I removed the check

File: dts/arm/st/h5/stm32h5.dtsi (hunk 7 above)
Comment (marwaiehm-st) [reply dts/arm/st/h5/stm32h5.dtsi:504]: No, i added for the both (TX and RX)

File: dts/arm/st/h5/stm32h5.dtsi (hunk 4 above)
Comment (marwaiehm-st) [reply dts/arm/st/h5/stm32h5.dtsi]: Yes , it should be used also in TX

File: drivers/i2s/i2s_ll_stm32.c [hunk 12]
```diff
@@ -349,9 +349,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
...
+				/*
+				 * Indicate that the transition to I2S_STATE_STOPPING
+				 * is triggered by DRAIN command
+				 */
+				stream->sample_tx_stop_drain = true;
+			} else {
+				stream->stream_disable(stream, dev);
+				stream->queue_drop(stream);
+				stream->state = I2S_STATE_READY;
+			}
```
Comment (erwango) [inline drivers/i2s/i2s_ll_stm32.c:361]: Non-RUNNING case already treated just above:
```
		if (stream->state != I2S_STATE_RUNNING) {
			irq_unlock(key);
			LOG_ERR("DRAIN trigger: invalid state");
			return -EIO;
		}
```

File: drivers/i2s/i2s_ll_stm32.c [hunk 13]
```diff
@@ -349,9 +349,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
 			LOG_ERR("DRAIN trigger: invalid state");
 			return -EIO;
 		}
-		stream->stream_disable(stream, dev);
-		stream->queue_drop(stream);
-		stream->state = I2S_STATE_READY;
+		if (dir == I2S_DIR_TX) {
```
Comment (erwango) [inline drivers/i2s/i2s_ll_stm32.c:349]: Should RX case be treated as documented: `	 * If the trigger is applied to the RX queue it has the same effect as * I2S_TRIGGER_STOP.` ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/i2s/i2s_ll_stm32.c (hunk 12 above)
Comment (mathieuchopstm) [reply drivers/i2s/i2s_ll_stm32.c:361]: +1

The check here should probably be something like `if (/* stream queue not empty || stream DMA busy */)` instead.

File: drivers/i2s/i2s_ll_stm32.c [hunk 14]
```diff
@@ -349,9 +359,30 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
...
+			if (stream->state == I2S_STATE_RUNNING) {
+				stream->state = I2S_STATE_STOPPING;
+				/*
+				 * Indicate that the transition to I2S_STATE_STOPPING
+				 * is triggered by DRAIN command
+				 */
+				stream->sample_tx_stop_drain = true;
+			} else if ((queue_is_empty(&stream->mem_block_queue) == false) ||
+						(LL_SPI_IsActiveFlag_TXC(cfg->i2s))) {
```
Comment (erwango) [inline drivers/i2s/i2s_ll_stm32.c:369]: An abstraction function will be required here to deal with F4/H5 variations:
```
static inline uint32_t ll_func_i2s_is_busy(I2S_TypeDef *i2s)
{
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_i2s)
	return LL_SPI_IsActiveFlag_TXC(i2s) == 0;
#else
	return LL_SPI_IsActiveFlag_BSY(i2s);
#endif /* st_stm32h7_i2s */
```

Comment (marwaiehm-st) [reply drivers/i2s/i2s_ll_stm32.c:369]: yes i forget to commit this modification :( i will do it now

File: drivers/i2s/i2s_ll_stm32.h [hunk 15]
```diff
@@ -58,4 +61,16 @@ struct i2s_stm32_data {
 	struct stream tx;
 };
 
+/* checks that DMA Tx packet is fully transmitted over the I2S */
+static inline uint32_t ll_func_i2s_dma_busy(SPI_TypeDef *i2s)
```
Comment (erwango) [inline drivers/i2s/i2s_ll_stm32.h:65]:
```suggestion
static inline uint32_t ll_func_i2s_dma_busy(I2S_TypeDef *i2s)
```
?

File: drivers/i2s/i2s_ll_stm32.h [hunk 16]
```diff
@@ -58,4 +61,16 @@ struct i2s_stm32_data {
...
+{
+#ifdef LL_SPI_SR_TXC
```
Comment (erwango) [inline drivers/i2s/i2s_ll_stm32.h:67]: #if #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_i2s)

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/i2s/i2s_ll_stm32.h (hunk 15 above)
Comment (marwaiehm-st) [reply drivers/i2s/i2s_ll_stm32.h:65]: The typdef of I2S does not exist in the hal_stm32, we used the spi type def in the structure i2s
```c
/* Device constant configuration parameters */
struct i2s_stm32_cfg {
	SPI_TypeDef *i2s;
	const struct stm32_pclken *pclken;
	size_t pclk_len;
	const struct pinctrl_dev_config *pcfg;
```

File: drivers/i2s/i2s_ll_stm32.c [hunk 17]
```diff
@@ -30,6 +30,16 @@ static unsigned int div_round_closest(uint32_t dividend, uint32_t divisor)
 	return (dividend + (divisor / 2U)) / divisor;
 }
 
+static bool queue_is_empty(struct ring_buf *rb)
+{
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:34]: This function is vulnerable to race conditions - while this isn't an issue ***now*** because all callers already suspend IRQs, it could become one in the future.
Please either:
* add a comment indicating this function must be called with IRQ suspended, or
* suspend interrupts for the duration of the function with an `irq_lock`/`irq_unlock` pair

Nit: this really should be a lock on the `struct ring_buf` instead of merely IRQ suspensions; this is acceptable(ish) because only one CPU at a time may interact with the DMA/I2S peripheral.

File: drivers/i2s/i2s_ll_stm32.c [hunk 18]
```diff
@@ -349,9 +359,30 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
...
+				 * Indicate that the transition to I2S_STATE_STOPPING
+				 * is triggered by DRAIN command
+				 */
+				stream->sample_tx_stop_drain = true;
+			} else if ((queue_is_empty(&stream->mem_block_queue) == false) ||
+						(ll_func_i2s_dma_busy(cfg->i2s))) {
+				stream->stream_disable(stream, dev);
+				stream->queue_drop(stream);
+				stream->state = I2S_STATE_READY;
+			} else
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:373]: Please add braces around `else` block.

File: drivers/i2s/i2s_ll_stm32.c [hunk 19]
```diff
@@ -349,9 +359,30 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
...
+				return -EIO;
+		}
+
+		if (dir == I2S_DIR_RX) {
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:377]: Please use `else if`.

Nit: replacing this code path with a `goto do_trigger_stop`, where `trigger_stop` is a label added in the `case I2S_TRIGGER_STOP` block above, could be cleaner, as it reduces code duplication and is more in line with the Zephyr API documentation:
```
If the [I2S_TRIGGER_DRAIN] trigger is applied to the RX queue it has the same effect as I2S_TRIGGER_STOP.
``` 

Nit 2: I2S_TRIGGER_STOP is currently broken - maybe a comment should be added to indicate this?

File: drivers/i2s/i2s_ll_stm32.c [hunk 20]
```diff
@@ -597,12 +639,41 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
...
+	 * Check if the tx stop command can be handled
+	 * if True, the tx stop command can be handled
+	 */
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		/*
+		 * Check if all tx samples have been completely handled
+		 * as stated in zephyr i2s specification, in case of DRAIN command
+		 * send all data in the transmit queue and stop the transmission.
+		 */
+		if (stream->sample_tx_block_sent == stream->sample_tx_block_dma_sent) {
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:658]: Nit: what if you replace this with `queue_is_empty`?

File: drivers/i2s/i2s_ll_stm32.h [hunk 21]
```diff
@@ -41,6 +41,9 @@ struct stream {
 	bool src_addr_increment;
 	bool dst_addr_increment;
 	uint8_t fifo_threshold;
+	uint8_t sample_tx_block_sent;
+	uint8_t sample_tx_block_dma_sent;
+	bool sample_tx_stop_drain;
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.h:46]: Nit: variable names are confusing, maybe names such as the following could be more appropriate?
* `uint8_t blocks_queued_by_api`
* `uint8_t blocks_sent_by_dma`
* `bool tx_stop_for_drain`

**Review by mathieuchopstm — CHANGES_REQUESTED** [review]
Mostly LGTM.
A few points to review:

File: drivers/i2s/i2s_ll_stm32.c (hunk 20 above)
Comment (marwaiehm-st) [reply drivers/i2s/i2s_ll_stm32.c]: I replace it with queue_is_empty , it didn't solve the last block problem but its more logical
 because the variable stream->sample_tx_block_sent it is initialized to 0 but never seems to be updated .

File: drivers/i2s/i2s_ll_stm32.c [hunk 22]
```diff
@@ -336,10 +350,21 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
...
+			stream->state = I2S_STATE_STOPPING;
+			/* indicate that the transition to I2S_STATE_STOPPING*/
+			/* is triggered by STOP command                      */
+			stream->tx_stop_for_drain = false;
+		} else if ((queue_is_empty(&stream->mem_block_queue) == false) ||
+					(ll_func_i2s_dma_busy(cfg->i2s))) {
+			stream->stream_disable(stream, dev);
+			stream->queue_drop(stream);
+			stream->state = I2S_STATE_READY;
+			stream->last_block = true;
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:364]: This doesn't comply with the API description which states:
```
   Stop the transmission / reception of data at the end of the current
   memory block. This trigger can be used in RUNNING state only and at
   first changes the interface state to STOPPING. When the current TX /
   RX block is transmitted / received the state is changed to READY.
   Subsequent START trigger will resume transmission / reception where
   it stopped.
```

I think the whole `if`/`else if`/`else` can be replaced with something like this: (also fixes the logic bug above)
```c
if (/* dma busy */) {
  stream->state = STOPPING
  stream->tx_stop_for_drain = false;
} else {
  stream->stream_disable(/* ... */); //i.e., stop the DMA
  //update state, switch to READY
}
```

_Make sure **not** to call `stream->queue_drop()`._

File: drivers/i2s/i2s_ll_stm32.c [hunk 23]
```diff
@@ -349,9 +374,29 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
 			LOG_ERR("DRAIN trigger: invalid state");
 			return -EIO;
 		}
-		stream->stream_disable(stream, dev);
-		stream->queue_drop(stream);
-		stream->state = I2S_STATE_READY;
+
+		if (dir == I2S_DIR_TX) {
+			if (stream->state == I2S_STATE_RUNNING) {
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:379]: Same logic bug as above. Similar pattern to above should work:

```c
if (/* dma busy */ || /* queue not empty */) {
   stream->state = STOPPING;
   stream->tx_stop_for_drain = true;
} else { //queue drained and no DMA op in progress
   //disable DMA
   //update state, switch to READY
}
```

File: drivers/i2s/i2s_ll_stm32.c [hunk 24]
```diff
@@ -603,6 +659,33 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
...
+		 * Check if all tx samples have been completely handled
+		 * as stated in zephyr i2s specification, in case of DRAIN command
+		 * send all data in the transmit queue and stop the transmission.
+		 */
+		if (queue_is_empty(&stream->mem_block_queue) == true) {
+			LOG_INF("All the data has been transferred");
+			stream->queue_drop(stream);
+			stream->state = I2S_STATE_READY;
+			goto tx_disable;
+		} else {
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:677]: Merge nested `else`+`if` into `else if` for improved readability.
Also add a comment like:
```c
/* else: DRAIN trigger -> continue TX normally until queue is empty */
```

File: drivers/i2s/i2s_ll_stm32.c [hunk 25]
```diff
@@ -603,6 +659,33 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
 		goto tx_disable;
 	}
 
+	/*
+	 * Check if the tx stop command can be handled
+	 * if True, the tx stop command can be handled
+	 */
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:665]: Nit: outdated/unclear comment?
```c
/* Check if we finished transferring one block and stopping is requested */
```

File: drivers/i2s/i2s_ll_stm32.c [hunk 26]
```diff
@@ -529,6 +574,16 @@ static void dma_rx_callback(const struct device *dma_dev, void *arg,
 
 	mblk_tmp = stream->mem_block;
 
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		if (stream->tx_stop_for_drain == false) {
+			/*
+			 * In case of STOP command, just stop the transmission
+			 * at the current. The transmission can be resumed
+			 */
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:582]: Comment refers to `transmission` but this is RX path.

File: drivers/i2s/i2s_ll_stm32.c [hunk 27]
```diff
@@ -529,6 +574,16 @@ static void dma_rx_callback(const struct device *dma_dev, void *arg,
 
 	mblk_tmp = stream->mem_block;
 
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		if (stream->tx_stop_for_drain == false) {
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:578]: There is no need to check this on RX because `I2S_TRIGGER_DRAIN` is redirected to `I2S_TRIGGER_STOP`. Ignoring it also removes the need to update it in `i2s_stm32_trigger`.

**Review by mathieuchopstm — CHANGES_REQUESTED** [review]
We're getting there! A few points to iron out:

File: drivers/i2s/i2s_ll_stm32.c [hunk 28]
```diff
@@ -603,6 +659,33 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
...
+	 */
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		/*
+		 * Check if all tx samples have been completely handled
+		 * as stated in zephyr i2s specification, in case of DRAIN command
+		 * send all data in the transmit queue and stop the transmission.
+		 */
+		if (queue_is_empty(&stream->mem_block_queue) == true) {
+			LOG_INF("All the data has been transferred");
+			stream->queue_drop(stream);
```
[the remark this answers is not in the document]
Comment (ajarmouni-st) [reply drivers/i2s/i2s_ll_stm32.c:674]: Also `LOG_INF` here should be `LOG_DBG`

File: drivers/i2s/i2s_ll_stm32.c [hunk 29]
```diff
@@ -349,9 +371,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
...
+		if ((queue_is_empty(&stream->mem_block_queue) == false) ||
+					(ll_func_i2s_dma_busy(cfg->i2s))) {
+			stream->state = I2S_STATE_STOPPING;
+			/*
+			 * Indicate that the transition to I2S_STATE_STOPPING
+			 * is triggered by DRAIN command
+			 */
+			stream->tx_stop_for_drain = true;
+		} else {
+			/* else: DRAIN trigger -> continue TX normally until queue is empty */
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:384]: misplaced comment

File: drivers/i2s/i2s_ll_stm32.c [hunk 30]
```diff
@@ -349,9 +371,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
 			LOG_ERR("DRAIN trigger: invalid state");
 			return -EIO;
 		}
-		stream->stream_disable(stream, dev);
-		stream->queue_drop(stream);
-		stream->state = I2S_STATE_READY;
+
```
Comment (mathieuchopstm) [inline drivers/i2s/i2s_ll_stm32.c:374]:
```c
if (dir == I2S_DIR_RX)
  goto trigger_stop;
```
is missing again

**Review by mathieuchopstm — APPROVED** [review]
Tiny non-blockers, otherwise LGTM.

Nit: `stream->last_block` seems to have become a dead variable, could used to be removed if so.

File: drivers/i2s/i2s_ll_stm32.h (hunk 15 above)
Comment (jeppenodgaard) [reply drivers/i2s/i2s_ll_stm32.h:63]: Why is this function not in the source file?

File: drivers/i2s/i2s_ll_stm32.h [hunk 31]
```diff
@@ -58,4 +59,16 @@ struct i2s_stm32_data {
 	struct stream tx;
 };
 
+/* checks that DMA Tx packet is fully transmitted over the I2S */
+static inline uint32_t ll_func_i2s_dma_busy(SPI_TypeDef *i2s)
```
Comment (jeppenodgaard) [inline drivers/i2s/i2s_ll_stm32.h:63]: Why is this function not in the source file?

Comment (erwango) [reply drivers/i2s/i2s_ll_stm32.h:63]: We're already doing this in spi driver for function which are only providing hw abstraction

Comment (erwango) [reply drivers/i2s/i2s_ll_stm32.h:63]: Improving driver file clarity. I agree that in current case, for one function, this might appear an excessive measure.
For spi driver case, there are lot of functions abstracted so it really made sense (this is all subjective, I agree).
Reusing the same scheme is for coherency sake across STM32 drivers.

**Review by mathieuchopstm — APPROVED** [review]
LGTM.

**Review by erwango — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Review by jeppenodgaard — APPROVED** [review]
